libethereum/Account.cpp

```diff
@@ -34,18 +34,24 @@ using namespace dev::eth::validation;
 
 namespace fs = boost::filesystem;
 
-void Account::setCode(bytes&& _code)
+void Account::setCode(bytes&& _code, u256 const& _version)
```

Comment:
Not related to your changes, but why do we have a setCode function rather than only allowing code to be set in the account object in a ctor since from what I understand smart contracts can't be redeployed at the same address?

Reply:
I think because contract creation has init code execution phase. First init code (provided in the contract deployment transaction or in CREATE/CREATE2 params) is executed, and it returns in the end another code, which becomes the code of new contract, saved into the state (with `setCode`).

So init code execution is performed in the context of the newly created contract, and so this new contract should already exist, but its code is empty until init code is finished.

Reply:
Ah okay, thanks for the clarification! Still not clear on this:

>  and it returns in the end another code, which becomes the code of new contract, saved into the state (with setCode).

What is this other code that is returned by init code execution, how does it differ from the code supplied in the contract creation tx?

Reply:
The code supplied in the contract creation tx (or in `CREATE`/`CREATE2` instructions) can return any data (via `RETURN` instruction). The rules are that the data returned from this code becomes the code of the newly created account.

In other words the code supplied in the contract creation tx is only constructor code (in solidity terms) and it _returns_ the code that becomes the actual contract code.